cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,363 @@
+
+CrossEntropyPair = NamedTuple('CrossEntropyPair', [('num_cycle', int),
+                                                   ('xeb_fidelity', float)])
+
+
+class CrossEntropyResult:
+    """Results from a cross-entropy benchmarking (XEB) experiment."""
+
+    def __init__(self, cross_entropy_pair: Sequence[CrossEntropyPair]):
+        """
+        Args:
+            cross_entropy_pair: A sequence of NamedTuples, each of which
+                contains two fields: num_cycle which returns the circuit
+                depth as the number of cycles and xeb_fidelity which returns
+                the XEB fidelity after the give cycle number.
+        """
+        self._data = cross_entropy_pair
+
+    @property
+    def data(self) -> Sequence[CrossEntropyPair]:
+        """Returns a sequence of NamedTuples, each of which contains a cycle
+        number and the corresponding XEB fidelity.
+        """
+        return self._data
+
+    def plot(self, **plot_kwargs: Any) -> None:
+        """Plots the average XEB fidelity vs the number of cycles.
+
+        Args:
+            **plot_kwargs: Arguments to be passed to matplotlib.pyplot.plot.
+        """
+        num_cycles = [d.num_cycle for d in self._data]
+        fidelities = [d.xeb_fidelity for d in self._data]
+        fig = plt.figure()
+        ax = plt.gca()
+        ax.set_ylim([0, 1.1])
+        plt.plot(num_cycles, fidelities, 'ro-', figure=fig, **plot_kwargs)
+        plt.xlabel('Number of Cycles', figure=fig)
+        plt.ylabel('XEB Fidelity', figure=fig)
+        fig.show()
+
+
+def cross_entropy_benchmarking(
+        sampler: sim.Sampler,
+        qubits: Sequence[ops.Qid],
+        *,
+        benchmark_ops: Sequence[Set[ops.OP_TREE]] = None,
+        num_circuits: int = 20,
+        repetitions: int = 1000,
+        cycles: Union[int, Iterable[int]] = range(2, 103, 10),
+        single_qubit_gates: List[List[ops.SingleQubitGate]] = None
+) -> CrossEntropyResult:
+    r"""Cross-entropy benchmarking (XEB) of multiple qubits.
+
+    A total of M random circuits are generated, each of which comprises N
+    layers where N = max(cycles) or cycles if a single value is specified for
+    the "cycles" parameter. Every layer contains randomly generated single-qubit
+    gates applied to each qubit, followed by a set of user-defined benchmarking
+    operations (e.g. a set of two-qubit gates).
+
+    Each circuit (circuit_m) from the M random circuits is further used to
+    generate a set of circuits {circuit_mn}, where circuit_mn is built from the
+    first n cycles of circuit_m. n spans all the values in cycles.
+
+    For each value n in cycles, the experiment performs the following:
+
+    1) Experimentally estimate the bit-string probabilities P^(meas, m)_|...00>,
+    P^(meas, m)_|...01>, P^(meas, m)_|...10>, P^(meas, m)_|...11>... at the
+    end of circuit_mn for all m. This is done by taking a total of R
+    projective measurements for each circuit_mn, histogram the collected
```

Comment:
Is histogram a verb? You might mean binning.

Reply:
Yes I meant binning. Changed the verb here.